ruma-common/src/push.rs

```diff
@@ -80,6 +85,97 @@ impl Ruleset {
             AnyPushRule::Sender(r) => self.sender.insert(r),
         }
     }
+
+    /// Get the push actions that apply to this event.
+    ///
+    /// Returns `None` if no push action applies.
+    ///
+    /// # Arguments
+    ///
+    /// * `event` - The raw JSON of a room message event.
+    /// * `context` - The context of the message and room at the time of the event.
+    pub fn get_actions<'a, T>(
+        &'a self,
+        event: &Raw<T>,
+        context: &PushConditionRoomCtx,
+    ) -> Option<impl Iterator<Item = &'a Action>> {
+        let event_map = &flatten_json(event);
+
+        for rule in self.override_.iter().filter(|r| r.enabled) {
+            if rule.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.content.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "content.body".to_string(),
+                pattern: rule.pattern.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.room.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "room_id".to_string(),
+                pattern: rule.rule_id.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.sender.iter().filter(|r| r.enabled) {
+            let condition = PushCondition::EventMatch {
+                key: "sender".to_string(),
+                pattern: rule.rule_id.clone(),
+            };
+
+            if condition.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+        for rule in self.underride.iter().filter(|r| r.enabled) {
+            if rule.applies(event_map, context) {
+                return Some(rule.actions.iter());
+            }
+        }
+
+        None
+    }
+}
+
+fn flatten_json<T>(json: T) -> BTreeMap<String, String>
+where
+    T: Serialize,
+{
+    let flat = &mut BTreeMap::new();
+
+    flatten_value(to_json_value(json).expect("Could not parse JSON"), flat, "".to_string());
```

Comment:
`to_json_value` is extremely unlikely to fail and I'm not sure "Could not parse JSON" if serialization fails is a helpful error message. I'd rather use `unwrap` here, or change the message to something like "Failed to serialize event".